python/tutorials/02-fused-softmax.py

```diff
@@ -81,7 +81,7 @@ def softmax_kernel(
     # Subtract maximum for numerical stability
     row_minus_max = row - tl.max(row, axis=0)
     # Note that exponentials in Triton are fast but approximate (i.e., think __expf in CUDA)
-    numerator = tl.exp(row_minus_max)
+    numerator = tl.libdevice.exp(row_minus_max)
```

Comment:
Error reported before this modification:

```
ptxas fatal   : Unresolved extern function '__nv_expf'
```

Reply:
I think we should keep tl.exp as a core function. We can map it to tl.libdevice.exp internally.

Reply:
Where will it report error if tl.exp? If in TritonGPUToLLVM we could just add an instantiation of the elementwise converters. As discussed with Keren we are going to keep both of the APIs for the moment.

Reply:
I'm trying to see if we miss some tests in test_core.py, but I noticed that tl.exp should be something that is already verified in https://github.com/openai/triton/blob/61f2ff98df81ad749cf050bfa430cc9312b9e720/python/tests/test_core.py#L467. So where am i missing?

Reply:
The error will be reported at ptxas phase.
@ben-zhang-609 What's your advice? Thanks

Reply:
For tl.exp, triton will map it to mlir::math::ExpOp, when translate to llvm ir with ptx as backend, the compiler will map it to __nv_expf.

On this whole path, triton has no ideal which external lib this instruction come from, where is the lib path.

For fixing this issue,  either I mapping these kind of functions to libdevice at frontend. 
Or always link libdevice when ptx is backend.

What you think?

Reply:
In theory, we should be able to know if libdevice is needed at the start of TritonGPUToLLVM, we don't need any additional information besides of the graph. Probably a op list is needed to tell if an op requires libdevice.

Reply:
only op list is not enough, the linker also need libdevise path.

Reply:
I don't quite follow, isn't libdevice something we should get from the cuda path?

Reply:
I'm pretty confused to. Triton already installs a `libdevice.bc.10` as part of the package. We should use this: https://github.com/openai/triton/blob/master/python/triton/language/libdevice.10.bc.

As for `exp`, we should have it be an alias for `libdevice.exp` to keep both APIs, but we need to make sure that we compile with `fast-math` by default.

Reply:
I get your both points. Let me do as following, link with libdevice by default when target is ptx, and using this default path https://github.com/openai/triton/blob/master/python/triton/language/libdevice.10.bc when user not specified.

Reply:
please check this PR: https://github.com/openai/triton/pull/848